### PR TITLE
Fix softmax op when the input shape is larger than INT32_MAX

### PR DESCRIPTION
**Pull request by sneaxiy**
### PR types
Bug fixes

### PR changes
OPs

### Describe
Fix softmax op when the input shape is large, i.e. `tensor.numel() >= std::numeric_limits<int32_t>::max()`.

### CODE REVIEW

File: paddle/phi/kernels/gpudnn/softmax_gpudnn.h [hunk 1]
```diff
@@ -758,11 +761,17 @@ void LaunchNormalSoftmaxBackward(const GPUContext& dev_ctx,
...
 
-static std::vector<int> GetSoftmaxTensorDims(const phi::DDim& dims,
-                                             const int axis) {
-  int dim = dims[axis];
-  int N = phi::funcs::SizeToAxis(axis, dims);
-  int D = phi::funcs::SizeOutAxis(axis, dims);
+template <typename T = int>
+static std::vector<T> GetSoftmaxTensorDims(const phi::DDim& dims,
+                                           const int axis) {
+  auto dim = static_cast<T>(dims[axis]);
```
Comment (FeixLiu) [inline paddle/phi/kernels/gpudnn/softmax_gpudnn.h:767]: Why not change the behavior of `SizeToAxis`, such as add a template? Probably some other ops will encounter same problems when calling `SizeToAxis`?

Comment (sneaxiy) [reply paddle/phi/kernels/gpudnn/softmax_gpudnn.h:767]: Great idea! Would fix soon in next commit.

Comment (sneaxiy) [reply paddle/phi/kernels/gpudnn/softmax_gpudnn.h:767]: Done.

**Review by FeixLiu — APPROVED** [review]
LGTM

**Pull request merged**